## <font color=red> Predict House Price Using Advanced Regression

In [ ]:
!pip install xgboost

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
pd.set_option("display.max_columns",100)

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge,Lasso
from sklearn.linear_model import ElasticNet
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import ExtraTreeRegressor
from sklearn.tree import DecisionTreeRegressor
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error,r2_score,mean_absolute_error


In [ ]:
def algo_test(x,y):
        #Bütün modelleri tanımlıyorum
        L=LinearRegression()
        R=Ridge()
        Lass=Lasso()
        E=ElasticNet()
        ETR=ExtraTreeRegressor()
        GBR=GradientBoostingRegressor()
        kn=KNeighborsRegressor()
        dt=DecisionTreeRegressor()
        xgb=XGBRegressor()
       
        
        algos=[L,R,Lass,E,ETR,GBR,kn,dt,xgb]
        algo_names=['Linear','Ridge','Lasso','ElasticNet','Extra Tree','Gradient Boosting','KNeighborsRegressor','Decision Tree','XGBRegressor']
        
        x_train, x_test, y_train, y_test=train_test_split(x,y,test_size=.2,random_state=42)
        
        r_squared= []
        rmse= []
        mae= []
        
        #Hata ve doğruluk oranlarını bir tablo haline getirmek için bir dataframe oluşturuyorum
        result=pd.DataFrame(columns=['R_Squared','RMSE','MAE'],index=algo_names)
        
        
        for algo in algos:
            p=algo.fit(x_train,y_train).predict(x_test)
            r_squared.append(r2_score(y_test,p))
            rmse.append(mean_squared_error(y_test,p)**.5)
            mae.append(mean_absolute_error(y_test,p))
        
            

        #result adlı tabloya doğruluk ve hata oranlarımı yerleştiriyorum
        result.R_Squared=r_squared
        result.RMSE=rmse
        result.MAE=mae
        
       #oluşturduğum result tablosunu doğruluk oranına (r2_score) göre sıralayıp dönüyor
        rtable=result.sort_values('R_Squared',ascending=False)
        return rtable

In [ ]:
df1=pd.read_csv('train.csv')

In [ ]:
df2=pd.read_csv('test.csv')

- Hedef olan sütunu direk y ye atıyorum.

In [ ]:
y=df1['SalePrice']

In [ ]:
abs(df1.corr()['SalePrice'].sort_values(ascending=False).head(22))

SalePrice       1.000000
OverallQual     0.790982
GrLivArea       0.708624
GarageCars      0.640409
GarageArea      0.623431
TotalBsmtSF     0.613581
1stFlrSF        0.605852
FullBath        0.560664
TotRmsAbvGrd    0.533723
YearBuilt       0.522897
YearRemodAdd    0.507101
GarageYrBlt     0.486362
MasVnrArea      0.477493
Fireplaces      0.466929
BsmtFinSF1      0.386420
LotFrontage     0.351799
WoodDeckSF      0.324413
2ndFlrSF        0.319334
OpenPorchSF     0.315856
HalfBath        0.284108
LotArea         0.263843
BsmtFullBath    0.227122
Name: SalePrice, dtype: float64

In [ ]:
df1.shape,df2.shape

((1460, 81), (1459, 80))

In [ ]:
df=df1.append(df2)

- df2 yani test verisinde saleprice sütunu yok eski çözümde o sütunu ekleyip sıfırla doldurmuştun sonrada tahmin ettiğin değerleri bu sütuna atmıştın ama ona hiç gerek yok en başya hedef sütuna salprice ı eşitledik artık silebiliriz böylece veriler birleştiğinde yani data da 1459 boş saleprice satırım olmayacak.

In [ ]:
del df["SalePrice"]

- Object olan ve sayı değerine sahip olan sütunları ayrı ayrı görmek istiyorum çünkü object olanlardaki verileri None ile dolduruken,diğerlerini 0 veya 1 ile dolduracağım.

In [ ]:
df[df.select_dtypes(include=['object']).columns]

MSZoning Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
0          RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1          RL   Pave   NaN      Reg         Lvl    AllPub       FR2       Gtl   
2          RL   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   
3          RL   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl   
4          RL   Pave   NaN      IR1         Lvl    AllPub       FR2       Gtl   
...       ...    ...   ...      ...         ...       ...       ...       ...   
1454       RM   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1455       RM   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1456       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1457       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl   
1458       RL   Pave   NaN      Reg         Lvl    AllPub    Inside       Mod   

     Neighborhood Condition1 Condition2 BldgType HouseStyle RoofStyle  \
0         CollgCr       Norm       Norm     1Fam     2Story     Gable   
1         Veenker      Feedr       Norm     1Fam     1Story     Gable   
2         CollgCr       Norm       Norm     1Fam     2Story     Gable   
3         Crawfor       Norm       Norm     1Fam     2Story     Gable   
4         NoRidge       Norm       Norm     1Fam     2Story     Gable   
...           ...        ...        ...      ...        ...       ...   
1454      MeadowV       Norm       Norm    Twnhs     2Story     Gable   
1455      MeadowV       Norm       Norm   TwnhsE     2Story     Gable   
1456      Mitchel       Norm       Norm     1Fam     1Story     Gable   
1457      Mitchel       Norm       Norm     1Fam     SFoyer     Gable   
1458      Mitchel       Norm       Norm     1Fam     2Story     Gable   

     RoofMatl Exterior1st Exterior2nd MasVnrType ExterQual ExterCond  \
0     CompShg     VinylSd     VinylSd    BrkFace        Gd        TA   
1     CompShg     MetalSd     MetalSd       None        TA        TA   
2     CompShg     VinylSd     VinylSd    BrkFace        Gd        TA   
3     CompShg     Wd Sdng     Wd Shng       None        TA        TA   
4     CompShg     VinylSd     VinylSd    BrkFace        Gd        TA   
...       ...         ...         ...        ...       ...       ...   
1454  CompShg     CemntBd     CmentBd       None        TA        TA   
1455  CompShg     CemntBd     CmentBd       None        TA        TA   
1456  CompShg     VinylSd     VinylSd       None        TA        TA   
1457  CompShg     HdBoard     Wd Shng       None        TA        TA   
1458  CompShg     HdBoard     HdBoard    BrkFace        TA        TA   

     Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2  \
0         PConc       Gd       TA           No          GLQ          Unf   
1        CBlock       Gd       TA           Gd          ALQ          Unf   
2         PConc       Gd       TA           Mn          GLQ          Unf   
3        BrkTil       TA       Gd           No          ALQ          Unf   
4         PConc       Gd       TA           Av          GLQ          Unf   
...         ...      ...      ...          ...          ...          ...   
1454     CBlock       TA       TA           No          Unf          Unf   
1455     CBlock       TA       TA           No          Rec          Unf   
1456     CBlock       TA       TA           No          ALQ          Unf   
1457      PConc       Gd       TA           Av          GLQ          Unf   
1458      PConc       Gd       TA           Av          LwQ          Unf   

     Heating HeatingQC CentralAir Electrical KitchenQual Functional  \
0       GasA        Ex          Y      SBrkr          Gd        Typ   
1       GasA        Ex          Y      SBrkr          TA        Typ   
2       GasA        Ex          Y      SBrkr          Gd        Typ   
3       GasA        Gd          Y      SBrkr          Gd        Typ   
4       GasA        Ex

In [ ]:
df[df.select_dtypes(include=[np.number]).columns]

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1454  2915         160         21.0     1936            4            7   
1455  2916         160         21.0     1894            4            5   
1456  2917          20        160.0    20000            5            7   
1457  2918          85         62.0    10441            5            5   
1458  2919          60         74.0     9627            7            5   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0          2003          2003       196.0       706.0         0.0      150.0   
1          1976          1976         0.0       978.0         0.0      284.0   
2          2001          2002       162.0       486.0         0.0      434.0   
3          1915          1970         0.0       216.0         0.0      540.0   
4          2000          2000       350.0       655.0         0.0      490.0   
...         ...           ...         ...         ...         ...        ...   
1454       1970          1970         0.0         0.0         0.0      546.0   
1455       1970          1970         0.0       252.0         0.0      294.0   
1456       1960          1996         0.0      1224.0         0.0        0.0   
1457       1992          1992         0.0       337.0         0.0      575.0   
1458       1993          1994        94.0       758.0         0.0      238.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0           856.0       856       854             0       1710           1.0   
1          1262.0      1262         0             0       1262           0.0   
2           920.0       920       866             0       1786           1.0   
3           756.0       961       756             0       1717           1.0   
4          1145.0      1145      1053             0       2198           1.0   
...           ...       ...       ...           ...        ...           ...   
1454        546.0       546       546             0       1092           0.0   
1455        546.0       546       546             0       1092           0.0   
1456       1224.0      1224         0             0       1224           1.0   
1457        912.0       970         0             0        970           0.0   
1458        996.0       996      1004             0       2000           0.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0              0.0         2         1             3             1   
1              1.0         2         0             3             1   
2              0.0         2         1             3             1   
3              0.0         1         0             3             1   
4              0.0         2         1             4             1   
...            ...       ...       ...           ...           ...   
1454           0.0         1         1             3             1   
1455           0.0         1         1             3             1   
1456           0.0         1         0             4             1   
1457           1.0         1         0             3             1   
1458           0.0         2         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0                8           0       2003.0         2.0       548.0   
1                6           1       1976.0         2.0       460.0   
2                6           1       2001.0         2.0       608.0   
3                7           1       1998.0         3.0       642.0   
4

- from sklearn.impute import SimpleImputer
- imputer = SimpleImputer(strategy="median")
- data3=imputer.fit_transform(data3)
- data3=pd.DataFrame(data3,columns=idx)

  -  Bu kodu yazarsan eğer her sütundaki boş verileri o sütunun median ı ile doldurmuş olursun bu normalde her sütunu doldurmakla uğraşmak yerine kolay yoldur ancak burada bunu yaparsan o özelliğe sahip olmayan evlere o özelliği vermiş olursun.Mesela havuzu yada bodrum katı olmayan evlere median ile bir değer atadığın an o evlerde bu özellikler varmış gibi olur bu doğru olmaz.

In [ ]:
empty=df.isnull().sum()
empty[empty>0]

MSZoning           4
LotFrontage      486
Alley           2721
Utilities          2
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
KitchenQual        1
Functional         2
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
GarageFinish     159
GarageCars         1
GarageArea         1
GarageQual       159
GarageCond       159
PoolQC          2909
Fence           2348
MiscFeature     2814
SaleType           1
dtype: int64

- Hangi sütunlarda boş veriler var ve ne kadar boş veri var görmek istiyorum.

In [ ]:
df.sample(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
946    947          80       RL         70.0     8163   Pave   NaN      Reg   
1092  1093          50       RL         60.0     8400   Pave   NaN      Reg   
201    202          20       RL         75.0    10125   Pave   NaN      Reg   
473   1934          60       RL         72.0     8702   Pave   NaN      IR1   
418    419          50       RL         60.0     8160   Pave   NaN      Reg   
613    614          20       RL         70.0     8402   Pave   NaN      Reg   
864   2325          20       RL         59.0    10936   Pave   NaN      IR1   
983   2444          50       RM        120.0     9000   Pave   NaN      Reg   
107   1568          60       RL          NaN    17082   Pave   NaN      IR1   
669    670          30       RL         80.0    11600   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
946          Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1092         Bnk    AllPub    Inside       Gtl        SWISU       Norm   
201          Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
473          Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
418          Lvl    AllPub    Inside       Gtl      Edwards       Norm   
613          Lvl    AllPub    Inside       Gtl      Mitchel      Feedr   
864          Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
983          Lvl    AllPub    Corner       Gtl      OldTown       Norm   
107          Low    AllPub   CulDSac       Mod      ClearCr       Norm   
669          Lvl    AllPub    Inside       Gtl      Crawfor       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
946        Norm     1Fam       SLvl            5            6       1959   
1092       Norm     1Fam     1.5Fin            6            5       1925   
201        Norm     1Fam     1Story            6            6       1977   
473        Norm     1Fam     2Story            6            5       1997   
418        Norm     1Fam     1.5Fin            5            6       1940   
613        Norm     1Fam     1Story            5            5       2007   
864        Norm     1Fam     1Story            7            5       2006   
983        Norm     1Fam     1.5Fin            5            8       1900   
107        Norm     1Fam     2Story            6            5       1978   
669        Norm     1Fam     1Story            4            5       1922   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
946           1959     Gable  CompShg     HdBoard     HdBoard    BrkFace   
1092          1950     Gable  CompShg     MetalSd     MetalSd       None   
201           1977     Gable  CompShg     Plywood     Plywood       None   
473           1998     Gable  CompShg     VinylSd     VinylSd       None   
418           1950     Gable  CompShg     MetalSd     MetalSd       None   
613           2007     Gable  CompShg     VinylSd     VinylSd       None   
864           2006     Gable  CompShg     VinylSd     VinylSd       None   
983           2006     Gable  CompShg      Stucco      Stucco       None   
107           1992     Gable  CompShg     VinylSd     VinylSd    BrkFace   
669           1950     Gable  CompShg     MetalSd     MetalSd       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
946        128.0        TA        Gd     CBlock       TA       TA   
1092         0.0        TA        TA      PConc       TA       TA   
201          0.0        TA        TA     CBlock       TA       TA   
473          0.0        TA        TA      PConc       TA       TA   
418          0.0        TA        TA     BrkTil       TA       TA   
613          0.0        TA        TA      PConc       Gd       TA   
864          0.0        Gd        TA      PConc       Gd       TA   
983          0.0        TA        TA     BrkTil       TA       Fa   
107        288.0        TA  

- Peki ben evlere olmayan özelliklerini kazandırmadan boş verileri nasıl dolduracağım??
  - Kategori olan sütunların boş verilerini None ile dolduracağım.
    - Allay,PoolQC,MiscFeature,Alley,Fence,FireplaceQu :mesela bu sütunlara çoğunda NaN yazıyor bu evlerde havuz,şömine,çit vb.olmadığını görüyoruz ve bu bilgileri ortalama bir değer ile doldurup şöminesi olmayan bir eve şöminesi var demek istemiyoruz bu sebeple bu category sütunlarının boş verilerini None ile dolduruyoruz. 

- df["PoolQC"] = df["PoolQC"].fillna("None")
- df["MiscFeature"] = df["MiscFeature"].fillna("None")
- df["Alley"] = df["Alley"].fillna("None")
- df["Fence"] = df["Fence"].fillna("None")
- df["FireplaceQu"] = df["FireplaceQu"].fillna("None")
  - Bu yolla da aşağıdaki yollada doldurabilirsin ikisinide öğren çünkü bir projede biri daha uygunken diğerinde diğeri daha uygun olabilir.Ama bir programcı aynı yada benzer satır kodları tekrar tekrar yazmaz onun bir kolay yolunu bulmaya çalışır.Aşağıdaki 2 satırlık kod gibi.

In [ ]:
for col in ('GarageType', 'GarageFinish', 'GarageQual', 'GarageCond','PoolQC','MiscFeature','Alley','Fence','FireplaceQu'):
    df[col] = df[col].fillna('None')

- LotFrontage:Evin caddeye olan uzaklığı diyebiliriz.Peki bunları nasıl dolduracağız?Burada ki boş verileri None ile doldurmayacağız çünkü sütun object değil.Ve biz şunu biliyoruz ki bu veriler boş bile olsa her evin caddeye belli bir uzaklığı vardır ve filmlerde görüyoruz ki çoğu ev belli bir düzen ile inşa edilmiş.
- Bu durumda biz boş olan bu ev-cadde arası uzaklıkları orta bir değer ile doldurabiliriz.

In [ ]:
df["LotFrontage"] = df.groupby("Neighborhood")["LotFrontage"].transform(lambda x: x.fillna(x.median()))

- lambda tek kullanımlık fonksiyondur tek satır ile tüm boş verileri lambda ile doldurabiliriz.

- Peki sıra geldi numeric olan sütunların boş verilerini nasıl dolduracağız.Bunu yapmanın en zekice yolu hiçbir eve ekstra özellik kazandırmadan o evde bu özelliğin olmadığını belirtmektir.Ve bunu da boş verilerin yerine 0 vererek yapıyoruz.Sonuçta garaj alanı boşsa biz bunu garajı yok şeklinde ele alabiliriz.Bu nedenle 0 vermemiz sorun olmaz.
- GarageYrBlt,GarageArea,GarageCars,BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, BsmtFullBath,BsmtHalfBath sütunlarında bodrum veya garaj yok şeklinde dolduracağız.

In [ ]:
for col in ('GarageYrBlt', 'GarageArea', 'GarageCars','BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF','TotalBsmtSF', 'BsmtFullBath', 'BsmtHalfBath'):
    df[col] = df[col].fillna(0)

- BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1,BsmtFinType2 bu sütunlar ise bodrumla ilgili kategorik özelliklere sahip olan sütunlardır.Daha önced deiğim gibi bunlarıda None ile dolduracağız.

In [ ]:
for col in ('BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2'):
    df[col] = df[col].fillna('None')

- MasVnrArea:duvar kağıdı alan 0 ile,MasVnrType:duvar kağıdı tipi ise None ile doldurulur.
- Not:Bazen sütunların ne ifade ettiğini anlayamayabilirsin(kısaltılmış şeklide yazıldığından vs.) bu durumlarda object ve numerical sütundan hangisi olduğuna bakıp ona göre None veya 0 ile doldurabilirsin.

In [ ]:
df["MasVnrType"] = df["MasVnrType"].fillna("None")
df["MasVnrArea"] = df["MasVnrArea"].fillna(0)

- MSZoning (The general zoning classification) :İmar , bir belediyenin veya başka bir hükümet kademesinin araziyi, her biri diğer bölgelerden farklı yeni gelişim için bir dizi düzenlemeye sahip olan bölgeler olarak adlandırılan alanlara böldüğü bir şehir planlama yöntemidir .Bu sütunu en yaygın olan değer ile dolduracağız.

In [ ]:
df['MSZoning']=df['MSZoning'].fillna(df['MSZoning'].mode()[0])

- [0] aynı sütundaki mode değerini bulmasını istediğimiz için yazdık.

- functional evin işlevselliğinin bir göstergesi olan bir sütundur.Kategorik bir sütundur ancak bunu None ile doldurmak eve bir fonksiyonu yok demek gibi bir şey olur bu yüzden böyle demektense bu sütunu da en yaygın olan değer ile dolduracağız.
- NOT:Sadece 1,2 gibi çok az sayıda boş veriye sahip sütunları o sütunun mode değeri ile doldurabiliriz.Mode fonk. güzel yanı hem kategorik hem sayısal sütunları doldurmak için kullanılabiliyor olmasıdır.

In [ ]:
for col in ('Functional','Electrical','KitchenQual','Exterior1st','Exterior2nd','SaleType','Utilities'):
    df[col] = df[col].fillna(df[col].mode()[0])

- MSSubClass:"Satışa dahil olan mesken tipi" olarak tanımlanır, sayısal olarak kodlanmıştır, ancak gerçekte kategorik bir değişkendir.(zipcode gibi sayılar barındırsada biz bu sütuna kategorik muamelesi yaparız.)
  - Burada dikkat etmemiz gereken en önemli şeylerden biri de bir sütun sayısal değerlere sahip diye o sütunun boş verilerini direk 0 ile dolduramayız.O sütun hakkında bir bilgimiz yoksa önce bir araştırma yapıp neyi temsil ettiğini anlamalıyız.
- Kategorik bir sütun olduğundan None ile dolduracağız.  

In [ ]:
df['MSSubClass'] = df['MSSubClass'].fillna("None")

In [ ]:
df.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MiscVal          0
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
Length: 80, dtype: int64

In [ ]:
empty=df.isnull().sum()
empty[empty>0]

Series([], dtype: int64)

- Artık hiç boş verimizin olmadığını hepsini uygun şekilde doldurduğumuzu görüyoruz.

- zipcode da yaptığımız gibi bir evin diğerine üstünlüğünü ifade etmeyen,sayısal verilerle ifade edilen,sütunları kategori türüne dönüştürüp dummie uygulayacağız.

In [ ]:
df['MSSubClass'] = df['MSSubClass'].apply(str)
df['OverallCond'] = df['OverallCond'].astype(str)
df['YrSold'] = df['YrSold'].astype(str)
df['MoSold'] = df['MoSold'].astype(str)

In [ ]:
df3=pd.get_dummies(df,drop_first=True)
df3.sample(15)

Id  LotFrontage  LotArea  OverallQual  YearBuilt  YearRemodAdd  \
1080  2541         64.0     8333            7       2006          2006   
638    639         67.0     8777            5       1910          1950   
403    404         93.0    12090            8       1998          1998   
438    439         40.0     4280            5       1913          2002   
1120  2581         65.0     6565            4       1957          1980   
1247  1248         74.0    12328            6       1976          1976   
538    539         72.0    11553            5       1968          1968   
870    871         60.0     6600            5       1962          1962   
117    118         74.0     8536            5       2006          2007   
830   2291        110.0    13440            8       2006          2007   
227    228         21.0     1869            6       1970          1970   
1284  1285         50.0     9638            6       1919          1990   
854    855        102.0    17920            5       1955          1974   
722    723         70.0     8120            4       1970          1970   
463    464         74.0    11988            6       1934          1995   

      MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  \
1080         0.0         0.0         0.0      738.0        738.0       738   
638          0.0         0.0         0.0      796.0        796.0       796   
403        650.0         0.0         0.0     1141.0       1141.0      1165   
438          0.0       365.0         0.0       75.0        440.0       694   
1120         0.0       967.0         0.0      106.0       1073.0      1073   
1247       335.0       539.0         0.0      473.0       1012.0      1034   
538        188.0       673.0         0.0      378.0       1051.0      1159   
870          0.0         0.0         0.0      894.0        894.0       894   
117          0.0         0.0         0.0     1125.0       1125.0      1125   
830        190.0         0.0         0.0     1108.0       1108.0      1148   
227        127.0       321.0         0.0      162.0        483.0       483   
1284         0.0         0.0         0.0      804.0        804.0      1699   
854          0.0       306.0      1085.0      372.0       1763.0      1779   
722          0.0       191.0         0.0      673.0        864.0       864   
463          0.0       326.0         0.0      389.0        715.0       849   

      2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  \
1080       753             0       1491           0.0           0.0         2   
638          0             0        796           0.0           0.0         1   
403       1098             0       2263           0.0           0.0         2   
438          0             0        694           0.0           0.0         1   
1120         0             0       1073           1.0           0.0         1   
1247         0             0       1034           1.0           0.0         1   
538          0             0       1159           0.0           0.0         1   
870          0             0        894           0.0           0.0         1   
117          0             0       1125           0.0           0.0         1   
830       1402             0       2550           0.0           0.0         2   
227        504             0        987           0.0           0.0         1   
1284       748             0       2447           0.0           0.0         2   
854          0             0       1779           1.0           0.0         1   
722          0             0        864           0.0           0.0         1   
463        811             0       1660           0.0           0.0         1   

      HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
1080         1             3             1             7           1   
638          0             2             1             4           0   
403          1             4             1            10     

- Id nin bir evin fiyatını belirlemesinde hiçbir etkisinin olmadığını biliyoruz bu nedenle tahmin ederken de dikkate almak istemediğimizden sütunu siliyoruz.

In [ ]:
del df3['Id']

In [ ]:
df1.shape,df2.shape

((1460, 81), (1459, 80))

In [ ]:
df12=df3[:1460]
df22=df3[1460:]

In [ ]:
algo_test(df12,y)

R_Squared          RMSE           MAE
Gradient Boosting     0.895958  28249.537173  17819.357670
Ridge                 0.876471  30781.630848  20142.826244
XGBRegressor          0.875698  30877.757974  17841.050768
Lasso                 0.840265  35003.172611  19448.488237
ElasticNet            0.833726  35712.418498  20981.323071
Decision Tree         0.824280  36712.837231  24415.945205
Extra Tree            0.729246  45571.600576  29438.880137
KNeighborsRegressor   0.699279  48027.365988  28554.471918
Linear               -0.039790  89305.796088  24526.087926

- En yüksek başarı oranına sahip olan olduğumuz model ile tahminlerimizi yapıp kaydedeceğiz.
  - Bunun için öncelikle boş bir dataframe (tablo) oluşturup feature engineering ile Id ve SalePrice sütunları açacağız.(Çünkü amacım hangi Id numarasına sahip olan ev ne kadarlık bir fiyata sahip olacak onu görmek.)Daha sonra da tahminlerimizi buradaki SalePrice sütununa atacağız.Ve bu dosyayı csv olarak kaydedeceğiz.

In [ ]:
df12_train, df12_test, y_train, y_test=train_test_split(df12,y,test_size=.2,random_state=42)
gb=GradientBoostingRegressor()
pred=gb.fit(df12_train,y_train).predict(df22)

In [ ]:
predictions=pd.DataFrame()
predictions['Id']=df2['Id']
predictions['SalePrice']=pred
predictions.to_csv('predictions.csv',index=False)
#index=False dememizin nedeni 0 dan başlayıp her satır sıralanan indexlerin kaydettiğim dosyada olmasını istememem.